File: p2.java
// Time Complexity : O(n*n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        // Sorting the array in the beginning
        Arrays.sort(nums);

        List<List<Integer>> result = new ArrayList<>();

        // Itearting through all the elements till we reach an element greater than 0 as
        // finding solution for element greater than 0 would not be possible

        for (int ind = 0; ind < nums.length - 1 && nums[ind] <= 0; ind++) {

            // Hashset for maintaining all the unique elements
            Set<Integer> set = new HashSet<>();
            // Iterating through the elements
            for (int i = ind + 1; i < nums.length; i++) {

                // Finding the element where all the three elements combined would result to 0
                int temp = 0 - nums[ind] - nums[i];

                // Adding the 3 elements
                if (set.contains(temp)) {
                    List<Integer> triplets = new ArrayList<>();
                    triplets.add(nums[ind]);
                    triplets.add(nums[i]);
                    triplets.add(temp);
                    result.add(triplets);
                    // To remove duplicates
                    while (i < nums.length - 1 && nums[i] == nums[i + 1])
                        i++;
                } else {
                    set.add(nums[i]);
                }

            }

            // To remove duplicated
            while (ind < nums.length - 1 && nums[ind] <= 0 && nums[ind] == nums[ind + 1])
                ind++;
        }

        return result;
    }

}

File: P2.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 12 11:37:48 2021

@author: jgdch
"""

#Time Compleixty: O(N**2), N is the length of input 
#Space complexity:O(1)
#All lc testcases passed

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if(len(nums)<3):
            return []
        nums.sort()
        res=[]
        for i in range(len(nums)-2):
            low=i+1
            high=len(nums)-1
            if(i>0 and nums[i-1]==nums[i]):
               continue
            while(low<high):
                sum=nums[i]+nums[low]+nums[high]
                if(sum==0):
                    res.append([nums[i],nums[low],nums[high]])
                    low+=1
                    high-=1
                    while(low < high and nums[low] == nums[low-1]) :
                        low+=1

                    while(low < high and nums[high] == nums[high + 1]) :
                        high-=1
                elif(sum<0):
                    low+=1
                else:
                    high-=1
        
        
        return res